ENT-4982 Database transaction handling with entity managers

lankydan opened the pull request:
## Summary

Improve the handling of database transactions when using
`withEntityManager` inside a flow.

Extra changes have been included to improve the safety and
correctness of Corda around handling database transactions.

This focuses on allowing flows to catch errors that occur inside an
entity manager and handle them accordingly.

Errors can be caught in two places:

- Inside `withEntityManager`
- Outside `withEntityManager`

Further changes have been included to ensure that transactions are
rolled back correctly.

## Catching errors inside `withEntityManager`

Errors caught inside `withEntityManager` require the flow to manually
`flush` the current session (the entity manager's individual session).
By manually flushing the session, a `try-catch` block can be placed
around the `flush` call, allowing possible exceptions to be caught.

Once an error is thrown from a call to `flush`, it is no longer possible
to use the same entity manager to trigger any database operations. The
only possible option is to rollback the changes from that session.
The flow can continue executing updates within the same session but they
will never be committed. What happens in this situation should be handled
by the flow. Explicitly restricting the scenario requires a lot of effort and
code. Instead, we should rely on the developer to control complex
workflows.

To continue updating the database after an error like this occurs, a new
`withEntityManager` block should be used (after catching the previous
error).

## Catching errors outside `withEntityManager`

Exceptions can be caught around `withEntityManager` blocks. This allows
errors to be handled in the same way as stated above, except the need to
manually `flush` the session is removed. `withEntityManager` will
automatically `flush` a session if it has not been marked for rollback
due to an earlier error.

A `try-catch` can then be placed around the whole of the
`withEntityManager` block, allowing the error to be caught while not
committing any changes to the underlying database transaction.

## Savepoints / Transactionality

To make `withEntityManager` blocks work like mini database transactions,
save points have been utilised. A new save point is created when opening
a `withEntityManager` block (along with a new session). It is then used
as a reference point to rollback to if the session errors and needs to
roll back. The save point is then released (independently from
completing successfully or failing).

Using save points means, that either all the statements inside the
entity manager are executed, or none of them are.

## Some implementation details

- A new session is created every time a entity manager is requested,
but this does not replace the flow's main underlying database session.
- `CordaPersistence.transaction` can now determine whether it needs
to execute its extra error handling code. This is needed to allow errors
escape `withEntityManager` blocks while allowing some of our exception
handling around subscribers (in `NodeVaultService`) to continue to work.

lankydan committed:
ENT-4982 Database transaction handling with entity managers

## Summary

Improve the handling of database transactions when using
`withEntityManager` inside a flow.

This focuses on allowing flows to catch errors that occur inside an
entity manager and handle them accordingly.

Errors can be caught in two places:

- Inside `withEntityManager`
- Outside `withEntityManager`

## Catching errors inside `withEntityManager`

Errors caught inside `withEntityManager` require the flow to manually
`flush` the current session (the entity manager's individual session).
By manually flushing the session, a `try-catch` block can be placed
around the `flush` call, allowing possible exceptions to be caught.

Once an error is thrown from a call to `flush`, it is no longer possible
to use the same entity manager to trigger any database operations. The
only possible option is to rollback the changes from that session.
Therefore if an errored session attempts to alter the database again, a
`RolledBackDatabaseSessionException` is thrown.

To continue updating the database after an error like this occurs, a new
`withEntityManager` block should be used (after catching the previous
error).

## Catching errors outside `withEntityManager`

Exceptions can be caught around `withEntityManager` blocks. This allows
errors to be handled in the same way as stated above, except the need to
manually `flush` the session is removed. `withEntityManager` will
automatically `flush` a session if it has not been marked for rollback
due to an earlier error.

A `try-catch` can then be placed around the whole of the
`withEntityManager` block, allowing the error to be caught while not
committing any changes to the underlying database transaction.

## Savepoints / Transactionality

To make `withEntityManager` blocks work like mini database transactions,
save points have been utilised. A new save point is created when opening
a `withEntityManager` block (along with a new session). It is then used
as a reference point to rollback to if the session errors and needs to
roll back. The save point is then released (independently from
completing successfully or failing).

Using save points means, that either all the statements inside the
entity manager are executed, or none of them are.

## Some implementation details

- A new session is created every time a entity manager is requested,
but this does not replace the flow's main underlying database session.
- `CordaPersistence.transaction` can now determine whether it needs
to execute its extra error handling code. This is needed to allow errors
escape `withEntityManager` blocks while allowing some of our exception
handling around subscribers (in `NodeVaultService`) to continue to work.

lankydan committed:
ENT-4982 Test tidy up

lankydan committed:
ENT-4982 Correct `transaction` parameter name

lankydan committed:
ENT-4982 Fix `AppendOnlyPersistentMapTest`

lankydan committed:
ENT-4982 Flush vault changes to underlying session

The `CashSelectionTest` highlighted a change in functionality due to
previous changes.

Saved transactions and states were not being flushed to the underlying
session inside of `NodeVaultService.notifyAll`. Every time a batch of
transaction updates are made to the database they are now also flushed.

This allows other operations inside the same database transaction to
observe the updates to the database that the previous calls made.

lankydan committed:
ENT-4982 Fix `VaultFlowTest`

This test was excluding the database error handling staff member. Due to
an extra flush an error was being thrown earlier (used to be thrown at
commit time). Being thrown earlier, stopped it from being caught by the
transition error staff member that kept it for observation. Letting the
database staff member handle it kept led to the same outcome and is the
expected result in a real node.

lankydan committed:
ENT-4982 Add extra `FlowEntityManagerTest` tests

- constraint violation inside a single entity manager throws
- entities are only committed when the flow suspends

lankydan committed:
ENT-4982 Add extra `FlowEntityManagerTest` test

- Check that a flow can continue interacting with other nodes after
catching an exception

- Tidy up the test some more

lankydan committed:
ENT-4982 Restrict `RestrictedEntityManager` more

- Restrict more functions inside of `RestrictedEntityManager`
- Allow `clear` to be called from `RestrictedEntityManager`
- Check if a session has been rolled back in more functions
- Add a test

lankydan committed:
ENT-4982 Closing a database transaction does not roll it back

## Description

Whether a database transaction is rolled back when closing its
connection/session depends on the database and its JDBC driver
(according to the internet anyway).

When an uncaught error propagates through the state machine, the current
database transaction is rolled back
(via `ActionExecutorImpl.executeRollbackTransaction`). How it did so,
was by just closing the transaction.

This did not seem to roll back the transaction, instead it was
committing uncommitted changes that were part of the current
transaction's session. This lead to data being persisted to the database
even though the flow failed. Furthermore, if the flow was to retry,
constraint violations could arise due to data that should not have been
committed actually residing in the database.

## Solution

The database transaction is now explicitly rolled back from two points:

- `ActionExecutorImpl.executeRollbackTransaction`
- `TransitionExecutorImpl.executeTransition`

All other instances of `DatabaseTransaction.close` were already rolling
back the transaction if an error occurred.

By explicitly rolling back, we no longer depend on the JDBC driver's
implementation to decide the behaviour.

lankydan committed:
ENT-4982 Fix detekt warning

lankydan committed:
ENT-4982 Add SQL entity manager tests

lankydan committed:
ENT-4982 Allow `find` to be called on session marked for rollback

`EntityManager.find` does not cause issues when a session is marked for
rollback. Therefore it should be allowed from `RestrictedEntityManager`,
even if the session is marked for rollback.

lankydan committed:
ENT-4982 Unrestrict functions

Allow a transaction marked for rollback to allow update statements to be
executed on it, even though none of the updates will actually go
through. Allow `EntityManager.transaction` to be called but restrict the
functions that can be called on it. This is to allow flows to control
what happens themselves, rather than enforcing it from Corda's
perspective.

lankydan committed:
ENT-4982 Add entity manager inside entity manager tests

lankydan commented:
run tests: slow

lankydan commented:
Whenever this is reviewed let me merge it :)

rick-r3 commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt:
```diff
@@ -196,7 +196,7 @@ class CordaPersistence(
         val transaction = contextTransaction
         try {
-            transaction.session.flush()
+//            transaction.session.flush()
```
?

We need the flush so that any database changes we have been making are reflected in the database connection we hand over.  What was the rationale for changing it?

We should be able to demonstrate this with some kind of test that records a transaction (with vault updates) and then tries to manually query the vault using this connection.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt:
```diff
@@ -247,6 +247,7 @@ class NodeVaultService(
         fun flushBatch(previouslySeen: Boolean) {
             val updates = makeUpdates(batch, statesToRecord, previouslySeen)
             processAndNotify(updates)
+            currentDBSession().flush()
```
Is this trying to compensate for the removed `flush` in the `jdbcSession` implementation ?

rick-r3 commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/DatabaseTransaction.kt:
```diff
@@ -73,6 +77,11 @@ class DatabaseTransaction(
         if (sessionDelegate.isInitialized()) {
+            // The [sessionDelegate] must be initialised otherwise calling [entityManager] will cause an exception
+            if(session.transaction.rollbackOnly) {
+                // I might be able to move this out if I am using the original session now?
```
I think it is still possible that _our_ session has encountered an issue and will be rolled back.

rick-r3 commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/DatabaseTransaction.kt:
```diff
@@ -124,4 +133,9 @@ class DatabaseTransaction(
+
+// should this just be an `IllegalStateException`?
+class RolledBackDatabaseSessionException : CordaRuntimeException(
+    "Further operations are being executed on a rolled back database session"
```
I think I would just go with "Attempt to commit database transaction marked for rollback".  Don't reveal details / concepts of sessions etc.

rick-r3 commented:
Can we try this against the different enterprise databases?  I think you might just need to prepare an Enterprise branch and make sure your tests will run as part of Integration Tests.  I want to make sure savepoints work.

lankydan committed:
Revert "ENT-4982 Flush vault changes to underlying session"

This reverts commit c9fafc84

lankydan committed:
ENT-4982 Re-add flush in `jdbcSession`

lankydan committed:
ENT-4982 Change `RolledBackDatabaseSessionException` message

lankydan committed:
ENT-4982 Test data saved from `rawUpdate` is seen in same transaction

lankydan committed:
ENT-4982 Always close entity manager sessions

lankydan committed:
ENT-4982 Savepoint fix for MSSQL and Oracle

SQL server and Oracle do not support `releaseSavepoint`. A work around
has been added to prevent errors from occurring.

lankydan committed:
ENT-4982 Tidy up

lankydan committed:
ENT-4982 tests

lankydan commented:
Replaced by https://github.com/corda/corda/pull/6187
